File: quicksort.js
/** @param {Number[]} arr */
// module.exports = function QuickSort (arr, left, right) {
function QuickSort (arr, left, right) {
    let pivot;
    let index;
    if (left < right) {
        pivot = right;
        index = Partition(arr, pivot, left, right);
        QuickSort(arr, left, index - 1);
        QuickSort(arr, index + 1, right);
    }
    return arr;
}
function Partition (arr, pivot, left, right){
	let index = left;
    for (let i = left; i < right; i++) {
        // console.log("in:", arr, "i:", i, "index", index, "right:", right);
        if (arr[i] < arr[pivot]) {
            console.log("in:", arr, "i:", i, "index", index, "right:", right, "pivot:", pivot);
            Swap(arr, i, index);
            index++;
        }
    }
    Swap(arr, right, index);
    // console.log("partition:", arr);
    return index;
}
function Swap (arr, i, j) {
    let tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
    console.log("swap:", arr, i, j);
}

// console.log(QuickSort([2,1,3,4,8,6,5], 0, 6));
console.log(QuickSort([5,4,6,2,1,3], 0, 5));

File: quicksort.c
#include <stdio.h>

void swap(int a, int b, int *arr) {
	int t = arr[a];
	arr[a] = arr[b];
	arr[b] = t;
}

void quickSort(int left,int right, int* data) {
	int pivot = left;
	int j = pivot;
	int i = left+1;

	if (left < right) {
		for (; i <= right; i++) {
			if (data[i] < data[pivot]) {
				j++;
				swap(j, i,data);
			}
		}
		swap(left,j, data);
		pivot = j;

		quickSort(left, pivot-1,data);
		quickSort(pivot+1, right, data);
	}

}

int main() {

	int arr[11] = { 5,4,3,2,1, 10, 11, 12 , 20, 14, 16};
	quickSort(0, 10, arr);
	// for (int i = 0; i < 11; i++) printf("%d ",arr[i]);

	return 0;
}